schema/any.go

```diff
@@ -69,16 +61,152 @@ func (a *AnySchema) ValidateCompatibility(typeOrData any) error {
+			}
+		}
+		if firstReflectKind == reflect.Invalid { // First item
+			firstReflectKind = reflectKind
+		} else if firstReflectKind != reflectKind {
+			return &ConstraintError{
+				Message: fmt.Sprintf(
+					"mismatched key types in map passed into 'any' type: %s != %s",
+					firstReflectKind, reflectKind),
+			}
+		}
+
+		// Validate value
+		err := a.ValidateCompatibility(value)
+		if err != nil {
+			return &ConstraintError{
+				Message: fmt.Sprintf("validation error while validating map item item %q of map for 'any' type (%s)", key, err.Error()),
+			}
+		}
+	}
+	return nil
+}
+
+//nolint:nestif
+func (a *AnySchema) validateAnyList(data []any) error {
+	if len(data) == 0 {
+		return nil // No items to check, and following code assumes non-empty list
+	}
+	// Test list items
+	for _, item := range data {
+		err := a.ValidateCompatibility(item)
+		if err != nil {
+			return &ConstraintError{
+				Message: fmt.Sprintf("validation error while validating list item of type `%T` in any type (%s)", item, err.Error()),
+			}
+		}
+	}
+	// validate that all list items are compatible with the first to make the list homogeneous.
+	firstItem := data[0]
+	firstItemType, firstValIsSchema := firstItem.(Type)
+	if firstValIsSchema {
+		for i := 1; i < len(data); i++ {
+			valToTest := data[i]
+			err := firstItemType.ValidateCompatibility(valToTest)
+			if err != nil {
+				return &ConstraintError{
+					Message: fmt.Sprintf(
+						"validation error while validating for homogeneous list item `%T` in any type %s",
+						valToTest, err.Error()),
+				}
+			}
+		}
+	} else {
+		// Loop through all items. Ensure they have the same type.
+		firstItemType := reflect.ValueOf(firstItem).Kind()
+		for i := 1; i < len(data); i++ {
+			valToTest := data[i]
+			typeToTest := reflect.ValueOf(valToTest).Kind()
+			if firstItemType != typeToTest {
+				// Not compatible or is a schema
+				schemaType, valIsSchema := valToTest.(Type)
+				if !valIsSchema {
+					return &ConstraintError{
+						Message: fmt.Sprintf(
+							"types do not match between list items passed for any type %T != %T; "+
+								"lists should have homogeneous types",
+							firstItem, valToTest),
+					}
+				} else {
+					err := schemaType.ValidateCompatibility(valToTest)
+					if err != nil {
+						return &ConstraintError{
+							Message: fmt.Sprintf(
+								"types do not match between list items passed for any type %s; "+
+									"lists should have homogeneous types",
+								err),
+						}
+					}
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (a *AnySchema) ValidateCompatibility(typeOrData any) error {
+	switch typeOrData := typeOrData.(type) {
+	case Type:
+		return a.validateSchemaCompatibility(typeOrData)
+	case map[string]any:
+		// Test individual values
+		for key, value := range typeOrData {
+			err := a.ValidateCompatibility(value)
+			if err != nil {
+				return &ConstraintError{
```

Comment:
Again, here and below, shouldn't we give the `%T` or equivalent to say *why* we're complaining?

Reply:
The wrapped error at the end `(%s)` should be enough.
`validation error while validating string-keyed map item item "b" of map for 'any' type (Validation failed: unsupported data type for 'any' type: schema_test.someStruct)`

To differentiate the errors I just added `int-keyed`, `string-keyed`, and `any-keyed` to clarify which map validation code failed.